qemu/tests/luks_image_over_qsd.py

```diff
@@ -0,0 +1,92 @@
+import socket
+
+from provider.qsd import QsdDaemonDev
+
+from virttest import error_context, env_process
+from virttest import qemu_storage
+from virttest import utils_numeric
+from virttest import data_dir
+from virttest.qemu_monitor import QMPCmdError
+
+
+# This decorator makes the test function aware of context strings
+@error_context.context_aware
+def run(test, params, env):
+    """
+        Export a luks image via QSD with NBD inet.
+        1. Create a luks image.
+        2. Start a QSD daemon, and export the luks image with NBD inet.
+        3. Check the info of the image over NBD.
+        4. Boot up a guest with the exported image as a data disk.
+        5. Check there is the data disk in the guest.
+        6. Connect to the QMP sock
+        6.1 Query the block exports
+        6.2 luks key management.
+        :param test: VT test object.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+    qsd = QsdDaemonDev("qsd1", params)
+    qsd.start_daemon()
+
+    luks_image_size = params['image_size_stg1']
+    luks_img_param = params.object_params('stg1')
+    luks_img = qemu_storage.QemuImg(luks_img_param, data_dir.get_data_dir(),
+                                    'stg1')
+    nbd_image_tag = params['nbd_image_tag']
+    nbd_image_params = params.object_params(nbd_image_tag)
+    localhost = socket.gethostname()
+    nbd_image_params['nbd_server'] = localhost if localhost else 'localhost'
+    qemu_img = qemu_storage.QemuImg(nbd_image_params, None, nbd_image_tag)
+    qemu_img.info()
+
+    params["start_vm"] = 'yes'
+
+    login_timeout = params.get_numeric("login_timeout", 360)
+    env_process.preprocess_vm(test, params, env, params.get("main_vm"))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    output = session.cmd_output("lsblk -nd | grep '[s|v]db'")
+    if not output:
+        test.fail("There is no the luks image in guest, check please")
+    elif luks_image_size not in output:
+        test.fail(
+            "The image size is not matched to the data disk in guest, "
+            "check please.")
```

Comment:
I am afraid we may hit error because the name of the luks disk in guest may not be sdb or vdb, it could be sda(we hit this issue before :)) , so we can use the following instead:

            disks = utils_disk.get_linux_disks(session, True)
            session.close()
            for _, attr in disks.items():
                if 'disk' in attr and luks_image_size in attr:
                    break
            else:
                test.fail("Failed to find the luks image in guest")
Note that the disks(image1 and stg1) size should be different, for we get disk via disk size

Another way to find out the disk in guest is the following:
set blk_extra_params_stg1 = "serial=TARGET_DISK0" in cfg
in py, 
         m = re.search(r"(serial|wwn)=(\w+)", image_params["blk_extra_params"], re.M)
         if m is not None:
             drive_path = utils_misc.get_linux_drive_path(session, m.group(2))
you can search get_linux_drive_path for the detailed use.